feature: Initial Pipelines

briancain opened the pull request:
This pull request introduces the `f-pipelines` feature branch into the `main` branch of Waypoint. I'm opening this PR with the intention of allowing the Waypoint team to take time to read through all of the changes and provide feedback before we merge it into the `main` branch to be iteratively improved on.

This feature branch houses a very basic MVP that introduces the baseline feature of running and configuring Pipelines for an application. There is still much work to be done, but this PR includes:

- Constructing pipelines of many steps with step dependencies in a `waypoint.hcl`
- Steps are backed by existing Waypoint job operations
- Running a pipeline queues steps as Waypoint jobs
- Steps support built-in reserved step plugins for `Build`, `Deploy`, `Release`, and `Up`.
- An "exec" step can be used for running custom commands
- CLI and APIs for syncing code, running a pipeline, and pipeline config introspection

There is still work to be done to make Pipelines useful in Waypoint, but this is a good milestone to getting the feature work merged into `main` so the team can continue to work on the feature.

Much of this branch has been reviewed already by various team members. Below represents most of the pull requests that went into working on pipelines:

## Testing Steps

The following example is with an example java app in Docker

#waypoint.hcl
```
project = "example-java"

pipeline "test" {
  step "build" {
    use "build" {
      disable_push = true
    }
  }
}

app "example-java" {
  build {
    use "pack" {
      builder = "gcr.io/buildpacks/builder:v1"
    }
  }
  deploy {
    use "docker" {
      service_port = 8080
    }
  }
}
```

`$ waypoint-dev config sync` 

(Set up remote file source in the UI)

`$ waypoint pipeline run test`

## Pipelines Work

@mitchellh  Pipelines work:
- Pipelines Data Model, State API, Upsert https://github.com/hashicorp/waypoint/pull/3283
- Pipeline Step Execution https://github.com/hashicorp/waypoint/pull/3320
- RunPipeline API https://github.com/hashicorp/waypoint/pull/3346
- Extract `internal/jobstream` from `internal/client` https://github.com/hashicorp/waypoint/pull/3373

@briancain  Pipelines work:
- Parsing pipeline and step configs as well as config validations https://github.com/hashicorp/waypoint/pull/3321
- Pipeline code config syncing: https://github.com/hashicorp/waypoint/pull/3334
- Job system bug with unimplemented WatchTasks: https://github.com/hashicorp/waypoint/pull/3344
- Pipeline CLI/API for `pipeline list`, `pipeline inspect`, and `GetPipeline` and `ListPipelines`: https://github.com/hashicorp/waypoint/pull/3353
- Syncing pipeline configs with `waypoint config sync`: https://github.com/hashicorp/waypoint/pull/3397
- `waypoint pipeline run` CLI implementation: https://github.com/hashicorp/waypoint/pull/3406
- `waypoint job get-stream` implementation: https://github.com/hashicorp/waypoint/pull/3410
- Adjusting the expire time for runner jobs: https://github.com/hashicorp/waypoint/pull/3460
  + Before, when a job as queued, it was expected that a runner would pick up a job right away. This isn't the case with pipelines, because we queue all jobs ahead of time. This PR adjusts the expire time to only come into play when a runner has accepted a job.
- Reserved built-in Steps for Pipelines (RFC WP-131)
  - Build: https://github.com/hashicorp/waypoint/pull/3459
  - Deploy: https://github.com/hashicorp/waypoint/pull/3470
  - Release: https://github.com/hashicorp/waypoint/pull/3491
  - Up: https://github.com/hashicorp/waypoint/pull/3493

mitchellh committed:
proto: pipeline data model

mitchellh committed:
pkg/serverstate: PipelinePut

mitchellh committed:
pkg/serverstate: PipelineGet

mitchellh committed:
pkg/serverstate: PipelineDelete

mitchellh committed:
pkg/server/ptypes: pipeline validation

mitchellh committed:
pkg/server/ptypes: validate step graph

mitchellh committed:
pkg/server: UpsertPipeline API

mitchellh committed:
internal/server: boltdb state should reuse pipeline validation

mitchellh committed:
proto

proto

proto

briancain committed:
internal/config: Introduce Pipelines and Steps for parsing configs

This commit introduces a new concept to a Waypoint HCL known as
Pipelines. A pipeline can have many Steps.

mitchellh committed:
pkg/server: implement task override and skipping

mitchellh committed:
internal/runner: pipeline step execution initial impl

Untested, most likely very broken.

mitchellh committed:
extract internal/jobstreamui

mitchellh committed:
internal/runner: stream child step UI events to the job UI

briancain committed:
internal/config: Update Pipeline to have multi-step stanzas

briancain committed:
internal/config: Validate pipeline stanzas

This commit adds some validation that pipelines must have at least one
step stanza, and includes a test to validate with example hcl configs.

briancain committed:
internal/config: Validate pipeline stanzas

This commit adds some validation that pipelines must have at least one
step stanza, and includes a test to validate with example hcl configs.

briancain committed:
internal/config: Update StepLabels to return all labels for all steps

mitchellh committed:
cli: job knows about about pipeline steps

mitchellh committed:
builtin/docker: don't error if error is nil!

briancain committed:
todo: Clean up some TODO comments

briancain committed:
internal/config: pipeline stanza label is Name

briancain committed:
pr feedback: remove unused references to pipelines

briancain committed:
internal/config: DependsOn is a slice of step names

briancain committed:
internal/config: Do a full pipeline and step decode all together

briancain committed:
Remove workspace and label scoped from Step struct

briancain committed:
internal/config: Introduce `hclStep` for raw Step

briancain committed:
internal/config: Configure step plugins in pipeline, not in operation

briancain committed:
internal/config: Move hclStep into config pipeline

briancain committed:
internal/core: Add ConfigSync for pipeline configs

This commit introduces a way for runners to initiate a pipeline config
sync. It loads the HCL eval context, evals the config, and translates the
parsed pipeline struct into a Proto Pipeline that is upserted to the server.

mitchellh committed:
internal/jobstream: log unknowns

briancain committed:
pr feedback: Simplify hclStep, make step label required for now

briancain committed:
internal/config: Panic on nil conifg when generating Pipeline Protos

briancain committed:
pr feedback: No dynamic funcs for pipeline configs

briancain committed:
internal/config: Set project owner on pipeline

briancain committed:
test: Update pipeline proto func test to operate on genericConfig

briancain committed:
internal/{cli,api}: Add Pipeline commands for run, list, and config sync

mitchellh committed:
proto

mitchellh committed:
pkg/server: add PipelienSteps

mitchellh committed:
pkg/server: job queue with ODR respects job dependencies

mitchellh committed:
proto

mitchellh committed:
server: RunPipeline

briancain committed:
server: Ensure primary ODR job gets added as dependency

Prior to this commit, there was a bug with ODR and generating the Stop
Task job. If a plugin did not implement WatchTask, the WatchTask job would fail,
which would invoke the StopTask job and break the task state machine while the
main job was still running. This commit fixes that behavior by ensuring
that if a source job has no overrides defined and is not expected to
skip the operation, the source id gets added as a dependent for the stop
task.

briancain committed:
maint: move task state logging into trace

mitchellh committed:
proto

mitchellh committed:
server: return job ids and map for run pipeline

briancain committed:
internal/server: Introduce ListPipelines API func

This commit introduces a simple API interface for listing any pipelines
for a given project from Waypoint.

briancain committed:
internal/cli: Add a list cli for Pipelines per project

briancain committed:
internal/cli: Add `waypoint pipeline inspect <pipeline-id>`

briancain committed:
Add rpc func doc for GetPipeline

briancain committed:
server/pipeline: Include root step name in get pipeline response

briancain committed:
server/proto: Update config sync pipeline response with a value

This commit updates the expected pipeline config sync response to
include a map of pipeline name to pipeline Ref Ids for each pipeline
that was synced in the request

briancain committed:
internal/{server,cli}: Update `waypoint config sync` to sync pipeline configs

mitchellh committed:
internal/jobstream: extracted for streaming job info

mitchellh committed:
internal/client: convert to jobstream package

mitchellh committed:
remove jobstreamui package

mitchellh committed:
avoid import cycles

briancain committed:
cli: Improve UX of config sync

Some simple UX improvements and better user feedback as to which
pipelines were synced, and where they get synced to.

briancain committed:
internal/core: Look up pipelines by Owner and Name

briancain committed:
internal: Update func doc comments

briancain committed:
Fix test comment

briancain committed:
internal/cli: Add `waypoint pipeline run` for running pipelines

Introduce a simple CLI for manually executing a pipeline by name and
pipeline owner.

briancain committed:
Show more job operations in task list

briancain committed:
Include UI for stream client, and stream download event

briancain committed:
internal/cli: Add pipeline id flag for running by id instead of name

This commit introduces the ability to run a pipeline in the CLI via its
specific UUID rather than a pipeline name and project owner.

briancain committed:
Fix panic if pipeline-id flag is used but no argument supplied to cli

briancain committed:
internal/cli: Implement `get-stream` for job CLI

This commit implements the get stream subcommand for the job CLI so that
users can attach to completed or currently running job streams.

briancain committed:
Add changelog

briancain committed:
internal/core: Look up pipeline id on upsert on server

Prior to this commit, the config sync frontend would look up if a
pipeline already existed by Owner and Name to see if it should reuse the
pipeline ID when a config sync happens. This was to prevent duplicate
pipelines being upserted with separate Ids. This commit changes that by
pushing back the pipeline lookup into the db side.

briancain committed:
test: Use sever package to ref TestRunner

briancain committed:
server/ptype: Validate that image is required for a step op

briancain committed:
internal/cli: Ensure workspace ref is set for job template

briancain committed:
Add some test hcl configs for exec step tests

briancain committed:
internal/config: Evaluate step exec plugin for PipelineProto

briancain committed:
Add more logging in service runner

briancain committed:
internal/cli: Add step name before job stream output

briancain committed:
internal/server: Add reserved built-in Pipeline Steps

This commit introduces the first built-in reserved pipeline step called
"build". It takes a Step config with a push switch (similar to the CLI)
and will queue a Build operation job based on the users Build stanza in
their application config

briancain committed:
Add note about setting a default step image URL

briancain committed:
Fix doc comment

briancain committed:
Add PipelineStep as possible job operation

briancain committed:
Extend expire time for queued jobs

This commit extends the expiry time for queued jobs from 1 minute to 60
minutes. The reason being that prior to pipelines, we only ever expected
queued jobs to be picked up immediately without any downstream jobs.
This no longer makes sense given that Pipelines can queue a LOT of jobs
at once and it is reasonable that they might wait a while before
executing.

A future todo here is to set a 60 second time out if a runner is running
and has accepted a job. There's no reason we should wait in that
instance.

briancain committed:
runner: Update pipeline step op to use job queue handler

briancain committed:
internal/server: Set expiry time on runner accept

Prior to this commit, Waypoint server only expected queued jobs to wait
for 60 seconds before a runner accepted the job and began executing it
from its creation time. With the introduction of pipelines, this no
longer makes sense because a pipeline will queue all jobs in a pipeline
right away. This means downstream jobs could be waiting around to be
executing for a while, much longer than 1 minute. This is expected.

Given that, this commit updates the server behavior to set a jobs expiry
time to something much longer, 60 minutes, and then when the runner goes
to accept the job, we update the expiry time to the original expected
behavior of 60 seconds, because if the runner has accepted a job to work
on, it should run that job fairly quickly like the original intended
behavior.

briancain committed:
Add func method doc

briancain committed:
pr feedback: remove initial expire time from run job

briancain committed:
pr feedback: Queue reserved step ops in RunPipeline service func

Rather than modify the runner operation for executing exec steps, this
commit updates the pipeline run behavior to queue a resereved build step
op when a RunPipeline request queues all of the jobs.

briancain committed:
internal/server: Support reserved pipeline step "deploy"

This commit introduces the deploy reserved built-in step for Pipelines.
It supports a Release bool, but intentionally does not implement release
yet as that will be coming in a follow-up commit/pull request to this
one.

It also introduces a change in the deploy operation in that if there was
no artifact set on the jobs Deploy operation, it looks up the latest
pushed artifact to use that. This change is important because pipelines
queues its jobs ahead of time all at once, which means if at any point a
Build step runs prior, we couldn't look up the artifact without it
existing. So the Deploy step will not set the artifact, making it nil,
and when the Deploy goes to run it will look up the latest.

briancain committed:
test: Update job test files with new impl state for db

briancain committed:
internal/jobstream: Show commit message in jobstream

briancain committed:
internal/{server,config}: Add Release reserved pipeline step

This commit introduces parsing the resereved release step. It also
introduces two new kinds of references: Ref_Artifact and Ref_Deployment.
This will be important because we need to let users reference a
Deployment either via latest (the default) or with a specific deployment
sequence number.

briancain committed:
internal/{server,runner}: Queue a Release job via a pipeline step

This commit updates the pipeline run service API func to queue a Release
job if requested from a pipeline step. Additionally, because pipelines
is expecting to support referencing "latest" deployments as well as
deployments by sequence numbers, the runner impl for running a Release
will now look up the latest deployment if the Job operation
configuration for a Deployment was not set i.e. nil.

briancain committed:
Fix doc comment typo

briancain committed:
internal/runner: Use `target` rather than the Deployment on the job operation

Prior to this commit, the operation release for a runner would operate
assuming that a jobs Release operation would always have a Deployment
proto attached to it. This changes in the pipelines world, because we
allow for Deployment Refs for queuing jobs ahead of time. Prior to even
pipelines, the runner release operation already was extracting out the
deployment target from the Release operation, but we'd still try to use
the deployment on the job proto instead.

This commit resolves that by making sure the release runner code uses
target.

briancain committed:
internal/{server/runner}: Ensure a Deployment is preloaded for Release

Prior to this commit, when a pipeline went to queue a Release job, it
didn't pre load the deployment proto with an artifact. The release
runner had a built-in assumption that all Deployments it received would
have a preloaded artifact attached. So this commit fixes that by listing
deployments via the service endpoint that preloads artifacts.

briancain committed:
fixup: Put back accidental deleted code comment

briancain committed:
internal/server: Implement Up reserved built-in step for pipelines

This commit introduces a built in Step for pipelines that executes a
full Build Deploy Release via the Up operation.

briancain committed:
config: Update pipeline validation to use new ValidationResult

briancain committed:
mergeup: Delete code that was accidentally added from rebase

briancain committed:
website auto-gen

briancain commented on pkg/server/proto/server.proto:
```diff
@@ -865,6 +892,24 @@ message Ref {
+
+      uint64 sequence = 2;
+    }
+  }
```
Here are the `Ref_Deployment` and `Ref_Artifact` proto messages.

briancain commented on pkg/server/singleprocess/service_job.go:
```diff
@@ -451,16 +508,6 @@ func (s *Service) onDemandRunnerStartJob(
-	}
-
-	job.ExpireTime = timestamppb.New(time.Now().Add(dur))
-
```
Note that we no longer expect runners to accept jobs ~60 seconds after a job is queued. This is because pipelines will queue all of the jobs ahead of time at once. In that case, a runner might not get to a job until much later. We instead set up an expire time of 60 seconds once the runner accepts a job.

briancain commented on pkg/server/singleprocess/service_pipeline.go:
```diff
@@ -0,0 +1,234 @@
+				// job to stepJobs
+
+				// Queue a release job too
+				log.Warn("Currently not queueing a release job yet....sry!!!")
```
I left this unimplemented because I'm not sure if it makes sense in the pipelines view. If someone wants to do a release, they'd just add a Release step....

briancain commented on internal/runner/operation_deploy.go:
```diff
@@ -24,6 +24,19 @@ func (r *Runner) executeDeployOp(
 		panic("operation not expected type")
 	}
 
+	if op.Deploy.Artifact == nil {
```
https://github.com/hashicorp/waypoint/issues/2884

xiaolin-ninja commented:
I updated the description to include steps for testing.

paladin-devops reviewed (commented):
Although I read through the earlier PRs, seeing this all come together is great! 😄 Still much to go through, but will be attempting to test this out and add more feedback.

briancain commented on .changelog/3410.txt:
````diff
@@ -0,0 +1,5 @@
+```release-note:improvement
````
This will come later! I'm unsure of when this PR will actually land in a release and what the actual release notes might look like so I left it out specifically for this PR.

briancain commented on internal/cli/pipeline_inspect.go:
```diff
@@ -0,0 +1,152 @@
+		},
+	}, terminal.WithInfoStyle())
+
+	// TODO(briancain): Use graphviz to build a pipeline graph and display in the terminal?
```
Yep, there's a lot of cool things we could do with the inspect command here! Something we should do after this gets merged though.

briancain commented on internal/cli/pipeline_inspect.go:
```diff
@@ -0,0 +1,152 @@
+		return 0
+	}
+
+	var owner string
```
Yeah, so in the context of a pipeline, a Pipeline owner is the project and pipeline name. In the future, it could be something else (we could make it a OneOf), but for now owner is which project a pipeline belongs to and its unique name.

briancain commented on internal/cli/pipeline_inspect.go:
```diff
@@ -0,0 +1,152 @@
+		}
+
+		if projectRef == nil {
+			c.ui.Output("You must specify a project with -project or be inside an existing project directory.\n"+c.Help(),
```
They are stored independently, but there is a field on Pipeline that references which project a pipeline refers to. I don't think it makes sense yet to have project-agnostic pipelines with the current Waypoint config model but maybe some day!

briancain commented:
Rebased off main.

paladin-devops commented on internal/config/pipeline.go:
```diff
@@ -0,0 +1,283 @@
+					return nil, diag
+				}
+
+				s.Kind = &pb.Pipeline_Step_Exec_{
```
⏳ For a "step plugin" to be supported in pipelines, does it need to be added to the server proto? I'm thinking about how that could be extracted to the plugin SDK so any plugin could be used here, if so, & where in this package/elsewhere in "core" would need to be changed to enable that.

briancain replied:
That is correct. We intentionally didn't do the work to support external step plugins yet because we were thinking we're pretty far off from users wanting custom step plugins. But this is a good catch/good eye, we should think about what it'd take to do this when we're ready for it!

paladin-devops commented on internal/config/pipeline.go:
```diff
@@ -0,0 +1,283 @@
+			// own Step Eval func that returns the step proto instead and does the
+			// switches there.
+			// NOTE(briancain): This is what you'd change to support future Step plugins
+			switch step.Use.Type {
```
⏳ I think in the future when running pipelines locally works (right now they work with remote ops only), we could also include the registry step here to run a push op. It's not necessary right now as it is since the build operation should also push in remote operations, so this is of course not blocking, but it came to mind when thinking about supporting each of the stanzas within `app`. 😄

briancain replied:
I do think there's value in allowing for locally run pipelines, especially while you're working on building them. Let's track that as a future enhancement for sure!

We can include a stand-alone registry step in follow-up work. I think it might be a bit tricky because we'll need to know what artifact we're actually pushing and or allow for referencing prior steps to push prior artifacts.

paladin-devops reviewed (commented):
My experience running pipelines in remote ops has been great! My `exec` steps ran successfully, and although my `build` step failed (due to issues in my env I'm working through), the `build` "job" was kicked off in a pipeline! 

Also the icons in my feedback have been from [this article's](https://www.netlify.com/blog/2020/03/05/feedback-ladders-how-we-encode-code-reviews-at-netlify/#-pebble-non-blocking-but-requires-future-action) "feedback ladder" for PR review, to help identify blocking/non-blocking/future consideration comments.

As this has been a long lived feature branch with several smaller PRs merged into it, I'm going to re-read some of the previous PRs to better understand how different parts of this one tie in together/work independently, but in my testing, all of the changes seem to be working together cohesively. 😄

briancain commented on internal/config/pipeline.go:
```diff
@@ -0,0 +1,283 @@
+				}
+
+				s.Kind = &pb.Pipeline_Step_Release_{
+					Release: &pb.Pipeline_Step_Release{
```
Interesting, these built-in step plugins match what options are available on the job operations parameters. I'll take another look at this!

briancain replied:
We took a look at this, and it's actually something special that the `waypoint release` CLI is doing and isn't related to the Release operation. If we want this behavior we should add it to the release operation, and then we get it in a release step for pipelines.

paladin-devops reviewed (approved):
![image](https://user-images.githubusercontent.com/83741749/180049256-eab7c670-4f9f-4cb8-8e15-bf3c5a3e7a2d.png)

catsby commented on website/content/commands/pipeline-run.mdx:
```diff
@@ -0,0 +1,44 @@
+
+#### Command Options
+
+- `-pipeline-id=<string>` - Run a pipeline by ID.
```
The usage block reads "run a pipeline by name" are ID and Name the same here?

briancain replied:
Name refers to the name from `pipeline "name" {}` where id refers to the uuid from its entry in the database, so they are different in this case. I think the default/expected case here is running by name and ID might be a special case, so I made ID a flag and name an argument.

catsby commented on website/data/commands-nav-data.json:
```diff
@@ -190,6 +190,18 @@
     "path": "plugin"
   },
+  {
+    "title": "pipeline inspect",
```
Should these commands go above "plugin" to keep it alpha-ordered?

briancain replied:
Oops, I think so yes!

catsby commented on internal/cli/main.go:
```diff
@@ -617,6 +617,35 @@ func Commands(
+				baseCommand: baseCommand,
+			}, nil
+		},
+		/*
```
Is this dead code or an oversight?

briancain replied:
I should delete this code!

catsby reviewed (commented):
Starting code review in chunks, probably only minor feedback anticipated as we've already reviewed this as a group internally

briancain commented on internal/core/project.go:
```diff
@@ -148,6 +150,27 @@ func NewProject(ctx context.Context, os ...Option) (*Project, error) {
+		if err != nil {
+			return nil, fmt.Errorf("error loading pipeline %q: %w", name, err)
+		}
+		if err := pipelineConfig.Validate(); err != nil {
```
I need to double check but I think this will already happen! I am fairly sure now that we do this validation in `PipelinePut`

briancain replied:
OK yeah, we do this validation already when a pipeline is upserted, but not at the parsing step. I think it's worth it to get a config validation error since it's not supported storing tin the db so I'll update this PR for that!

briancain replied:
Handled in [d7391fd](https://github.com/hashicorp/waypoint/pull/3506/commits/d7391fde08a39605212327ffdc142ba89991efb2)!

briancain commented on internal/client/job.go:
```diff
@@ -43,6 +41,15 @@ func (c *Project) job() *pb.Job {
+// used; it is dangerous because it doesn't perform many validation steps.
+// In almost all cases, callers should use a more focused function such as
+// Build or Deploy, or write a new one.
+func (c *Project) DoJobDangerously(ctx context.Context, job *pb.Job) (*pb.Job_Result, error) {
```
I know this was adding while Mitchell was working on pipeline step execution. Not sure if the intention was to use this eventually, but it does look like nothing calls this at the moment.

briancain committed:
maint: Remove commented out code

briancain committed:
maint: Fix code doc comment for Pipeline core init func

briancain committed:
maint: Spelling

briancain committed:
website: move pipeline nav links to alpha order

briancain commented on internal/config/testdata/pipelines/pipeline_multi_step.hcl:
```diff
@@ -0,0 +1,42 @@
+  step "one" {
+    image_url = "example.com/second"
+
+    use "test/exec" {
```
Aha, so I looked into why this wasn't causing a failure. I use this config specifically for the HCL parsing part of `config/pipelines`, where the actual `use "label"` validation for steps really happens over in `PipelineProtos()`. I removed the namespacing since it doesn't make sense for the config test this was being used for.

briancain committed:
test: remove use label namespacing for test

briancain committed:
internal/config: Validate that there are no duplicate names

This commit ensures at the validation step that Waypoint does not allow
two pipeline stanzas with the same name label.

briancain committed:
pr feedback: Simplify if-statement

briancain commented on internal/cli/pipeline_inspect.go:
```diff
@@ -0,0 +1,152 @@
+		return 1
+	} else {
+		pipelineId = c.args[0]
+	}
```
Handled in [371c758](https://github.com/hashicorp/waypoint/pull/3506/commits/371c7588f1f07be627971f67a9294ea2ffe7a35a)!

xiaolin-ninja commented:
Runner targeting validation parsing doesn't seem to work with pipelines.

`waypoint.hcl`:
```
project = "example-java"

runner {
  profile="testing"
}

pipeline "test" {
  step "build" {
    use "build" {
      disable_push = true
    }
  }
```
Error message:
```
» Running pipeline "test" for application "example-java"
✓ Pipeline "test" has started running. Attempting to read job stream sequentially in order

✓ Pipeline "test" has started running. Attempting to read job stream sequentially in order

» Executing Step "build"
  Reading job stream (jobId: 01G8VA93YRXAY5GQE82ME89X3Z)...

  Performing this operation on a remote runner with id "testing"

» Cloning data from Git
  URL: https://github.com/hashicorp/waypoint-examples
  Ref: xx/test-branch

» Downloading from Git
  Git Commit: fff3f1b636e82a30e169e061c3a86b2a59624c1f
   Timestamp: 2022-07-25 18:35:12 +0000 UTC
     Message: why
  
! 1 error occurred:
  	* rpc error: code = Unknown desc =
  /var/folders/jn/l867j07d5jg973xlrg5ysvbh0000gn/T/waypoint3112285330/docker/java/waypoint.hcl:7,1-9:
  Unsupported block type; Blocks of type "pipeline" are not expected here.
  ```

briancain commented:
For what it's worth I can't reproduce the config validation error 🤔

briancain commented:
I cherry-picked the integration test fix. I'll do a proper rebase off `main` when it's time to merge.

briancain committed:
internal/config: Fix config.Validate() var returns
